### DISCUSSION TITLE
Can flake-parts help with sharing code with `nixosModules/darwinModules` exports?

### DISCUSSION THREAD

**Discussion by z0al** [hercules-ci/flake-parts#265] [Q&A] [answered]
Given the following flake:

```nix
{
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";

    flake-parts.url = "github:hercules-ci/flake-parts";
  };

  outputs =
    { self
    , nixpkgs
    , flake-parts
    , ...
    }@inputs:
    let 
      # Q: How to ensure this is passed to nixosModules/darwinModules?
      lib = nixpkgs.lib.extend
        (final: prev: (import ./lib final));
    in 
    flake-parts.lib.mkFlake { inherit inputs; } {
      systems = [
        ...
      ];

      flake = {
        nixosModules.default = import ./module.nix;
        darwinModules.default = import ./darwin-module.nix;
      };
    };
}
```

How do I pass my custom lib to `nixosModules.default`?

I tried `import ./module.nix { inherit lib;}` but then it asks for `config` and other parameters I don't know how to provide.

Alternatively: what's the easiest way to share code (read custom helpers) between these exports in my case?

**Comment by roberth** [comment, marked as answer]
I think you're looking for [`importApply`](https://flake.parts/define-module-in-separate-file.html?highlight=importApply#importapply)

You can use it to pass anything to a module file.
Basically your module file will be a function to a module, arity 2:

```nix
{ self, customThing }:
{ config, withSystem, ... }:
{
  # module attrs
}
```

I would recommend to use a name that's not `lib` for anything other than the Nixpkgs `lib`, but some will disagree with that :)

> `import ./module.nix`

`import` will throw away the file name, and the module system won't know where it came from, leading to worse error messages.
`importApply` solves that problem.
If you did't need to pass anything custom, it's better to do `nixosModules.default = ./module.nix;` etc. Also applies to `imports`, options of type `deferredModule`, and anywhere a module can be passed.

**Reply by z0al** [reply]
> Basically your module file will be a function to a module

Oops, I thought I had a look at `importApply` before but somehow missed the point about that part so ofc it didn't work 🤦🏻‍♂️ 

Thanks a lot.

**Reply by z0al** [reply]
I should be able to use `importApply` even in the nested context, right? meaning if top-level nixosModule imports other sub-modules via `imports = [...]` that also need access to `myLib`

**Reply by roberth** [reply]
Yes, it returns a module, that can be used in those places.
